File: laravel/app/Http/Controllers/Api/Players/PlayerController.php
<?php

namespace App\Http\Controllers\Api\Players;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use App\Models\Players\Player;
use App\Models\User;

class PlayerController extends Controller
{
    public function index(): JsonResponse
    {
        return response()->json(Player::all());
    }

    public function store(Request $request): JsonResponse
    {
        $validated = $request->validate([
            'display_name' => 'required|string|max:50',
            'rank' => 'required|string|in:Bronze,Silver,Gold,Platinum,Diamond,Master,Grandmaster|max:20',
            'rating' => 'required|integer',
            'peak_rating' => 'required|integer',
            'bio' => 'nullable|string|max:200',
            'country_code' => 'nullable|string|max:2',
            'avatar_url' => 'nullable|string|url|max:200',
            'preferred_format' => 'nullable|string|in:Standard,Extended,Legacy,Vintage,Commander,Draft|max:20',
            'is_verified' => 'required|boolean',
            'created_at' => 'required|date',
            'last_active_at' => 'nullable|date',
            'user_id' => 'nullable|exists:users,id',
        ]);
        $item = Player::create($validated);
        $item->validateRules();

        return response()->json($item, 201);
    }

    public function show(Player $player): JsonResponse
    {
        return response()->json($player);
    }

    public function update(Request $request, Player $player): JsonResponse
    {
        $validated = $request->validate([
            'display_name' => 'sometimes|nullable|string|max:50',
            'rank' => 'sometimes|nullable|string|max:20',
            'rating' => 'sometimes|nullable|integer',
            'peak_rating' => 'sometimes|nullable|integer',
            'bio' => 'sometimes|nullable|string|max:200',
            'country_code' => 'sometimes|nullable|string|max:2',
            'avatar_url' => 'sometimes|nullable|string|url|max:200',
            'preferred_format' => 'sometimes|nullable|string|max:20',
            'is_verified' => 'sometimes|nullable|boolean',
            'created_at' => 'sometimes|nullable|date',
            'last_active_at' => 'sometimes|nullable|date',
            'user_id' => 'sometimes|nullable|exists:users,id',
        ]);
        $player->update($validated);
        $player->validateRules();

        return response()->json($player);
    }

    public function destroy(Player $player): JsonResponse
    {
        $player->delete();
        return response()->json(null, 204);
    }
    public function promote(Request $request, Player $player): JsonResponse
    {
        $result = $player->promote();
        $player->save();
        return response()->json(['result' => $result]);
    }

    public function demote(Request $request, Player $player): JsonResponse
    {
        $result = $player->demote();
        $player->save();
        return response()->json(['result' => $result]);
    }

    public function recordWin(Request $request, Player $player): JsonResponse
    {
        $player->recordWin();
        $player->save();
        return response()->json(null, 204);
    }

    public function recordLoss(Request $request, Player $player): JsonResponse
    {
        $player->recordLoss();
        $player->save();
        return response()->json(null, 204);
    }

    public function winRate(Request $request, Player $player): JsonResponse
    {
        $result = $player->winRate();
        $player->save();
        return response()->json(['result' => $result]);
    }

    public function verify(Request $request, Player $player): JsonResponse
    {
        $player->verify();
        $player->save();
        return response()->json(null, 204);
    }

    public function updateRating(Request $request, Player $player): JsonResponse
    {
        $delta = $request->input('delta');
        $player->updateRating($delta);
        $player->save();
        return response()->json(null, 204);
    }
}

File: phoenix/lib/cards_project_web/controllers/players/player_controller.ex
defmodule CardsProjectWeb.Players.PlayerController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Players
  alias CardsProject.Players.Player

  def index(conn, _params) do
    players = Players.list_players()
    json(conn, Enum.map(players, &serialize_player/1))
  end

  def show(conn, %{"id" => id}) do
    player = Players.get_player!(id)
    json(conn, serialize_player(player))
  end

  def create(conn, params) do
    case Players.create_player(params) do
      {:ok, player} ->
        conn
        |> put_status(:created)
        |> json(serialize_player(player))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def update(conn, %{"id" => id} = params) do
    player = Players.get_player!(id)
    case Players.update_player(player, params) do
      {:ok, player} ->
        json(conn, serialize_player(player))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  def delete(conn, %{"id" => id}) do
    player = Players.get_player!(id)
    Players.delete_player(player)
    send_resp(conn, :no_content, "")
  end

  # POST /api/players/{id}/promote
  def promote(conn, %{"id" => id}) do
    result = Players.player_promote_behavior(id)
    json(conn, %{result: result})
  end

  # POST /api/players/{id}/demote
  def demote(conn, %{"id" => id}) do
    result = Players.player_demote_behavior(id)
    json(conn, %{result: result})
  end

  # POST /api/players/{id}/win
  def record_win(conn, %{"id" => id}) do
    Players.player_record_win_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/players/{id}/loss
  def record_loss(conn, %{"id" => id}) do
    Players.player_record_loss_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # GET /api/players/{id}/win-rate
  def win_rate(conn, %{"id" => id}) do
    result = Players.player_win_rate_behavior(id)
    json(conn, %{result: result})
  end

  # POST /api/players/{id}/verify
  def verify(conn, %{"id" => id}) do
    Players.player_verify_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # PATCH /api/players/{id}/rating
  def update_rating(conn, %{"id" => id} = params) do
    delta = Map.get(params, "delta")
    Players.player_update_rating_behavior(id, delta)
    send_resp(conn, :no_content, "")
  end

  defp serialize_player(%Player{} = record) do
    Map.take(record, [:id, :display_name, :rank, :rating, :peak_rating, :bio, :country_code, :avatar_url, :preferred_format, :is_verified, :created_at, :last_active_at, :user_id, :season_stats_id])
  end

  defp format_errors(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end
end
